Start organizing output data with protocol buffers

dabreegster opened the pull request:
This takes the internal Rust representation of populations, households, MSOAs, etc that's built from all the raw data, and expresses it using protocol buffers. This PR includes the .proto, incomplete code to write the proto files from Rust, and sample Python code to print the protos as JSON (an easy way to debug / understand the data).

Main thing I'd like feedback on is the .proto:

1) Is any of the info we're including **extraneous** -- only useful to ASPICS and nobody else?
2) Are we **missing** any per-person or per-household data that we actually have in the raw input? We can include it in the .proto and plumb through if so.

Next steps on my end are to finish writing the proto files from Rust, then write the Python code to transform a proto file into a snapshot for ASPICS. At that point, we can split into 2 git repositories -- SPC and ASPICS. The snapshot conversion code will go in the ASPICS repo, leaving the SPC repo pure Rust (plus protocol buffer bindings and sample code in different languages).

dabreegster commented on init/src/synthpop.proto:
```diff
@@ -0,0 +1,110 @@
+syntax = "proto3";
+package synthpop;
+
+message Population {
```
I'll add many more comments here, and we can also write some higher-level docs with diagrams to describe the structure of the data

dabreegster commented on init/src/synthpop.proto:
```diff
@@ -0,0 +1,110 @@
+
+    uint32 age_years = 6;
+
+    Health health = 7;
```
I thought it'd be nice to organize things by category a bit -- health attributes fit nicely together. Maybe once we add in some more fields from the tus files (sex, income?), we could have a "Demographics" category and put age and SIC there too.

HSalat replied:
agreed

dabreegster commented on init/src/synthpop.proto:
```diff
@@ -0,0 +1,110 @@
+
+message Health {
+    Obesity obesity = 1;
+    // TODO Unclear what the values mean
```
@HSalat you sent a doc that had lots of details about all of these, but it's buried in Slack somewhere... Was it a google doc or overleaf, do you remember?

HSalat replied:
it's on overleaf. Here's. a copy:

idp: individual identifier as MSOA11CD_XXXXXX, where XXXXXX is the rank when the data are sorted by increasing hid, then decreasing age
MSOA11CD: Middle-layer Super Output Area as defined for the 2011 census where the individual lives
hid: household identifier, includes communal establishments

Identifiers for other data sets:
pid: 2011 Census
pid_tus: Time Use Survey 2015
pid_hse: Health Survey for England 2017

General information:
sex: 0 female, 1 male
age: in years
origin: 1 White, 2 Black, 3 Asian, 4 Mixed, 5 Other

Working and social characteristics:
nssec5: National Statistics Socio-economic classification:
    1. Higher managerial, administrative and professional occupations
    2. Intermediate occupations
    3. Small employers and own account workers
    4. Lower supervisory and technical occupations
    5. Semi-routine and routine occupations
    0. Never worked and long-term unemployed
soc2010: Previous version of the Standard Occupational Classification -- see: https://www.ons.gov.uk/methodology/classificationsandstandards/standardoccupationalclassificationsoc/soc2010
sic1d07: Standard Industrial Classification of Economic Activities 2007, 1st layer (number corresponding to the letter in alphabetical order) -- see: https://www.ons.gov.uk/methodology/classificationsandstandards/ukstandardindustrialclassificationofeconomicactivities
sic2d07: Standard Industrial Classification of Economic Activities 2007, 2nd layer -- see: https://www.ons.gov.uk/methodology/classificationsandstandards/ukstandardindustrialclassificationofeconomicactivities

Proportion of 24h spent doing different daily activities:
punknown + pwork + pschool + pshop + pservices + pleisure + pescort + ptransport  = pnothome
phome + pworkhome = phometot
pnothome + phometot = 1

Lockdown 'essential worker' status -- source: London in Lockdown: Batty et al. https://arxiv.org/abs/2011.07165:
keyworkercasa: -> we should change that to bool

Health data:
cvd: has a cardio-vascular disease (0 or 1)
diabetes: has diabetes (0 or 1)
bloodpressure: has high blood pressure (0 or 1)
BMIvg6: Body Mass Index:
    Not applicable
    Underweight: less than 18.5
    Normal: 18.5 to less than 25
    Overweight: 25 to less than 30
    Obese I: 30 to less than 35
    Obese II: 35 to less than 40
    Obese III: 40 or more
BMI_healthier: same as above, but everyone from Overweight to Obese III has been moved 1 rank higher in the above list (healthier population assumption)

Likelihoods compared to other individuals (i.e. not directly a probability) -- source: see Events WIP paper:
ESport: attending a live sports event
ERubgy: attending a live rugby event
EConcertF: attending a live concert primarily targeting a young female audience
EConcertM: attending a live concert primarily targeting a young male audience
EConcertFS: attending a live concert primarily targeting a middle-aged female audience
EConcertMS: attending a live concert primarily targeting a middle-aged male audience
EMuseum: visiting a museum
Geolocation columns:
lng: Longitude of the MSOA11CD centroid
lat:: Latitude of the MSOA11CD centroid

dabreegster commented on init/src/synthpop.proto:
```diff
@@ -0,0 +1,110 @@
+    // The Standard Industry Classification for where this person works. 0 means none.
+    uint64 sic1d07 = 5;
+
+    uint32 age_years = 6;
```
The SPC output is just meant to represent a population at a single point of time. If a downstream model (like ASPICS) wanted to model changes over time, this would be a possible approach.

Relatedly, ASPICS uses lockdown multipliers that have real dates associated. That data isn't plumbed through here yet. If and when we do, we'll have to be careful to represent dates properly

HSalat replied:
and this single point of time has to be 2020 as a matter of fact. It would be great if someone wants to dig into the problem of growing the population every year. SPENSER is supposed to be able to do that already.

dabreegster commented:
OK, I'm going to merge this in and continue with smaller commits. I'll open PRs for anything big / needing feedback. This PR now:

1) The Rust SPC pipeline emits a protobuf file
2) A new python script reads that proto, and converts to the snapshot format used in ASPICS
3) I removed the Rust snapshot conversion code and also a half-started version of ASPICS written in Rust. (In preparation for a clean SPC repo)

Everything works end-to-end; the snapshot produced from proto->python code is basically the same as the one produced by the earlier Rust converter. But of course we're not done. Some of my next steps:

- split the repository. I'll remove the SPC / Rust code from this one, and start a new SPC repo without any of the python
- continue fighting Mac issues for the ASPICS repo
- Handle lockdown stuff properly (we need to plumb through the actual date, so downstream people know when the multipliers apply to in actual time)
- Plumb through more fields described by Hadrien's doc
- Score a huge perf boost by managing MSOA strings more intelligently
- Add new study areas, and document how to do that
- Set up github action test automation for both repos
- Explore flatbuffers and some others, since protobuf has obnoxious limitations
